src/main/kotlin/fi/metatavu/noheva/api/DevicesApiImpl.kt

```diff
@@ -0,0 +1,155 @@
+
+        val device = deviceController.findDevice(id = deviceId)
+            ?: return createNotFound("Device $deviceId not found")
+
+        val foundExhibitionDevices = exhibitionDeviceController.listByDevice(device = device)
+
+        if (foundExhibitionDevices.isNotEmpty()) {
+            val exhibitionDeviceIds = foundExhibitionDevices.map { it.id }.joinToString()
+            return createBadRequest("Cannot delete device $deviceId because it's in use by exhibition devices $exhibitionDeviceIds")
+        }
+
+        deviceController.deleteDevice(device)
+
+        return createNoContent()
+    }
+
+    override fun findDevice(deviceId: UUID): Response {
+        loggedUserId ?: return createUnauthorized("Unauthorized")
+
+        val foundDevice = deviceController.findDevice(id = deviceId)
+            ?: return createNotFound("Device $deviceId not found")
+
+        return createOk(deviceTranslator.translate(foundDevice))
+    }
+
+    override fun getDeviceKey(deviceId: UUID): Response {
+        val foundDevice = deviceController.findDevice(id = deviceId)
+            ?: return createNotFound("Device $deviceId not found")
+
+        if (foundDevice.approvalStatus == DeviceApprovalStatus.APPROVED) {
+            val keyPair = cryptoController.generateRsaKeyPair()
+                ?: return createInternalServerError("Couldn't create keypair")
+
+            deviceController.storeDeviceKey(device = foundDevice, key = keyPair.public)
+
+            return createOk(DeviceKey(key = cryptoController.getPrivateKeyBase64(keyPair.private)))
+        }
+
+        return createForbidden(FORBIDDEN)
+    }
+
+    override fun listDevices(status: DeviceStatus?, approvalStatus: DeviceApprovalStatus?): Response {
+        loggedUserId ?: return createUnauthorized("Unauthorized")
+
+        val foundDevices = deviceController.listDevices(status = status, approvalStatus = approvalStatus)
+
+        return createOk(deviceTranslator.translate(foundDevices))
+    }
+
+    override fun updateDevice(deviceId: UUID, device: Device): Response {
+        val userId = loggedUserId ?: return createUnauthorized("Unauthorized")
+
+        val foundDevice = deviceController.findDevice(id = deviceId)
+            ?: return createNotFound("Device $deviceId not found")
+
+        var foundDeviceModel: DeviceModel? = null
+
+        if (device.deviceModelId != null) {
+            foundDeviceModel = deviceModelController.findDeviceModelById(device.deviceModelId)
+                ?: return createBadRequest("Device model ${device.deviceModelId} not found")
+        }
+
+        foundDevice.approvalStatus.let {
+            when (it) {
+                DeviceApprovalStatus.PENDING ->
+                    if (device.approvalStatus != it && device.approvalStatus != DeviceApprovalStatus.APPROVED) {
+                        return createBadRequest("Cannot change approval status")
+                    }
+                DeviceApprovalStatus.APPROVED ->
```

Comment:
Why do we allow changing approval status from approved to pending?

Reply:
I thought that just in case someone would want to revert the approval